### PR TITLE
feat: Implement conversations API

### PR DESCRIPTION
**Pull request by buibaoanh**
## Summary by CodeRabbit

* **New Features**
  * Conversations API: full CRUD for conversations and items with pagination, metadata filtering, and preserved IDs; route exposed in gateway.
  * Storage: default in-memory storage plus SQL-backed storage with multi-dialect support (Greptime, MySQL, Postgres, SQLite) and public storage interfaces/types.
  * Public schemas: rich item and conversation schemas/types for messages, files, images, function calls, reasoning, and list responses.

* **Tests**
  * Extensive end-to-end and storage test suites covering lifecycle, pagination, filtering, mounts, dialects, and edge cases.

* **Docs**
  * README updated with Conversations and storage usage, examples, and migration notes.

### CODE REVIEW

File: src/endpoints/conversations/handler.ts [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+        result = await create(ctx, signalLevel);
+      } else {
+        throw new GatewayError("Method Not Allowed", 405);
+      }
+    }
+
+    // GET/POST/DELETE /conversations/{id} (Conversation Instance)
+    else if (len === 2) {
+      const conversationId = segments[1] as string;
+      logger.debug(`[conversations] resolved conversation ID: ${conversationId}`);
```
Comment (heiwen) [inline src/endpoints/conversations/handler.ts]: Unnecessary debug statement, nothing was really resolved. This has been logged already in the initial .info log for the request.

File: src/endpoints/conversations/handler.ts [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+  }
+  addSpanEvent("hebo.request.deserialized");
+
+  const parsed = ConversationCreateBodySchema.safeParse(body);
+  if (!parsed.success) {
+    throw new GatewayError("Invalid Request", 400, undefined, parsed.error);
+  }
+  addSpanEvent("hebo.request.parsed");
+
+  const conversation = await withSpan("storage.createConversation", () =>
```
Comment (heiwen) [inline src/endpoints/conversations/handler.ts]: Please don't create additional spans, they are just noise / additional data / costs.

File: src/endpoints/conversations/storage/memory.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  async updateConversation(
+    id: string,
+    params: { metadata: Record<string, unknown> },
+  ): Promise<Conversation> {
+    const conversation = this.conversations.get(id);
+    if (!conversation) {
+      throw new Error(`Conversation not found: ${id}`);
+    }
+
+    const updated = {
```
Comment (heiwen) [inline src/endpoints/conversations/storage/memory.ts]: Why create a new object instead of modifying the existing object?

File: src/endpoints/conversations/storage/memory.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    const now = Math.floor(Date.now() / 1000);
+    const newItems: ConversationItem[] = items.map((item) => {
+      const conversationItem: ConversationItem = {
+        id: `item_${crypto.randomUUID()}`,
+        object: "conversation.item",
+        created_at: now,
+        ...item,
+      } as ConversationItem;
```
Comment (heiwen) [inline src/endpoints/conversations/storage/memory.ts]: Interface Design: Same comment as for creating new conversations. This seems to be logic that should be the same for all storage adapters. If that's the case, it should be done in the handler.

File: src/endpoints/conversations/handler.ts [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+      body = await ctx.request.json();
+    }
+  } catch {
+    throw new GatewayError("Invalid JSON", 400);
+  }
+  addSpanEvent("hebo.request.deserialized");
+
+  const parsed = ConversationCreateBodySchema.safeParse(body);
+  if (!parsed.success) {
+    throw new GatewayError("Invalid Request", 400, undefined, parsed.error);
```
Comment (heiwen) [inline src/endpoints/conversations/handler.ts]: We are using `z.preffityError(parsed.error)` for the error messages (also applies to other messages).

File: src/endpoints/conversations/storage/memory.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      throw new Error(`Conversation not found: ${conversationId}`);
+    }
+
+    existing.push(...items);
+    return items;
+  }
+
+  // eslint-disable-next-line require-await
+  async getItem(conversationId: string, itemId: string): Promise<ConversationItem | undefined> {
+    return this.items.get(conversationId)?.find((item) => item.id === itemId);
```
Comment (heiwen) [inline src/endpoints/conversations/storage/memory.ts]: Would it be more efficient to store this as a Map instead of an Array, to avoid O(n) of `find()` and use O(n) of `get()` instead?

File: src/endpoints/conversations/storage/memory.ts [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    let result = existing;
+
+    if (order === "desc") result = result.toReversed();
+
+    if (after) {
+      const i = result.findIndex((item) => item.id === after);
+      if (i !== -1) result = result.slice(i + 1);
+    }
+
+    return Promise.resolve(result.slice(0, limit));
```
Comment (heiwen) [inline src/endpoints/conversations/storage/memory.ts]: Is limit = 0 (meaning no limit) a valid value?

File: src/endpoints/conversations/storage/memory.ts [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    return Promise.resolve(this.conversations.get(conversationId));
+  }
+
+  listItems(conversationId: string, params?: ListItemsParams): Promise<ConversationItem[]> {
+    const existing = this.items.get(conversationId);
+    if (!existing) {
+      return Promise.reject(new Error(`Conversation not found: ${conversationId}`));
+    }
+
+    const { after, order = "desc", limit = 20 } = params ?? {};
```
Comment (heiwen) [inline src/endpoints/conversations/storage/memory.ts]: You would want to infer defaults in the handler, not in the storage abstraction layer.

File: src/endpoints/conversations/utils.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * Creates a new ConversationItem object from input data with generated ID and timestamp.
+ */
+export function createConversationItem(input: ConversationItemInput): ConversationItem {
+  return {
+    id: `item_${crypto.randomUUID()}`,
+    object: "conversation.item",
+    created_at: Math.floor(Date.now() / 1000),
+    ...input,
```
Comment (heiwen) [inline src/endpoints/conversations/utils.ts]: Avoid spreads. Just modify the original object.

File: src/endpoints/conversations/storage/memory.ts (hunk 6 above)
Comment (heiwen) [reply src/endpoints/conversations/storage/memory.ts]: Yes, concern is about performance. Maps in JS/TS also preserve the order of insert.

File: src/endpoints/conversations/storage/memory.ts [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+    let result = existing;
+
+    if (order === "desc") result = result.toReversed();
```
[the remark this answers is not in the document]
Comment (buibaoanh) [reply src/endpoints/conversations/storage/memory.ts]: `bun run lint --fix` prefer `toReversed()` over `reserve()`

File: tsconfig.json [hunk 11]
```diff
@@ -1,8 +1,8 @@
 {
   "compilerOptions": {
     // Environment setup & latest features
-    "lib": ["ES2022", "DOM", "DOM.Iterable"],
-    "target": "ES2022",
+    "lib": ["ES2023", "DOM", "DOM.Iterable"],
+    "target": "ES2023",
```
Comment (heiwen) [inline tsconfig.json]: Understand why you need to change `lib` to `2023`, but target should remain `2022` so it transpiles down.

File: src/endpoints/conversations/utils.ts (hunk 9 above)
Comment (heiwen) [reply src/endpoints/conversations/utils.ts]: Think about it this way: If you don't have the helper function, that's exactly what you would do. You own the caller, so you can control side-effect impact. It's good practice, because it avoids using more memory.

File: src/endpoints/conversations/schema.ts [hunk 12]
```diff
@@ -0,0 +1,211 @@
+import * as z from "zod";
+
+// --- Common ---
+
+export const MetadataSchema = z
+  .record(z.string().max(64), z.union([z.string().max(512), z.number(), z.boolean()]))
+  .refine((m) => Object.keys(m).length <= 16, {
```
Comment (heiwen) [inline src/endpoints/conversations/schema.ts]: We don't do `.refine()` level validation in the other schema's. Let's not do that right now here either. Creates unnecessary overhead.

File: src/endpoints/conversations/schema.ts [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+
+export const ResponseInputFileSchema = z
+  .object({
+    type: z.literal("input_file"),
+    filename: z.string().optional(),
+    file_data: z.string().nullable().optional(),
+    file_id: z.string().nullable().optional(),
+    file_url: z.string().nullable().optional(),
+  })
+  .refine((data) => data.file_data || data.file_id || data.file_url, {
```
Comment (heiwen) [inline src/endpoints/conversations/schema.ts]: We don't do .refine() level validation in the other schema's. Let's not do that right now here either. Creates unnecessary overhead.

File: src/endpoints/conversations/storage/memory.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    this.items = new LRUCache<string, Map<string, ConversationItem>>({
+      maxSize,
+      sizeCalculation: (items) => Math.max(1, this.estimateSize(items)),
+      dispose: (_value, key) => {
+        this.conversations.delete(key);
+      },
+    });
+  }
+
+  private estimateSize(obj: unknown): number {
```
Comment (heiwen) [inline src/endpoints/conversations/storage/memory.ts]: Faster and more accurate version (according to Codex):

```
private estimateSize(root: unknown): number {
  let total = 0;
  const stack: unknown[] = [root];

  while (stack.length) {
    const obj = stack.pop();
    if (obj == null) continue;

    const t = typeof obj;
    if (t === "string") {
      total += (obj as string).length * 2;
      continue;
    }
    if (t !== "object") continue;

    if (ArrayBuffer.isView(obj)) {
      total += (obj as ArrayBufferView).byteLength;
      continue;
    }

    if (Array.isArray(obj)) {
      const arr = obj as unknown[];
      for (let i = 0, n = arr.length; i < n; i++) stack.push(arr[i]);
      continue;
    }

    if (obj instanceof Map) {
      for (const [k, v] of obj as Map<unknown, unknown>) {
        stack.push(k);
        stack.push(v);
      }
      continue;
    }

    const rec = obj as Record<string, unknown>;
    for (const k in rec) stack.push(rec[k]);
  }

  return total;
}
```

File: src/endpoints/conversations/schema.ts [hunk 15]
```diff
@@ -0,0 +1,226 @@
+import * as z from "zod";
+
+// --- Common ---
+
+export const MetadataSchema = z.record(
+  z.string().max(64),
+  z.union([z.string().max(512), z.number(), z.boolean()]),
+);
+
+export const ItemStatusSchema = z.enum(["in_progress", "completed", "incomplete"]);
```
Comment (heiwen) [inline src/endpoints/conversations/schema.ts]: Style: Move this to where it is used

File: src/endpoints/conversations/schema.ts [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+
+export const ResponseSummaryTextSchema = z.object({
+  type: z.literal("summary_text"),
+  text: z.string(),
+});
+
+export const ResponseReasoningTextSchema = z.object({
+  type: z.literal("reasoning_text"),
+  text: z.string(),
+});
```
Comment (heiwen) [inline src/endpoints/conversations/schema.ts]: Style: I'm trying to understand this schema and have to jump up and down. Please move things close to the block where they are actually used first.

File: src/endpoints/conversations/schema.ts [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+export type ConversationUpdateBody = z.infer<typeof ConversationUpdateBodySchema>;
+
+export const ConversationDeletedSchema = z.object({
+  id: z.string(),
+  deleted: z.boolean(),
+  object: z.literal("conversation.deleted"),
+});
+export type ConversationDeleted = z.infer<typeof ConversationDeletedSchema>;
+
+// Conversation: Stored
```
Comment (heiwen) [inline src/endpoints/conversations/schema.ts]: What does `Stored` mean?

File: src/endpoints/conversations/schema.ts [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+  status: ItemStatusSchema.optional(),
+});
+
+const FunctionCallOutputSchema = z.object({
+  type: z.literal("function_call_output"),
+  id: z.string().optional(),
+  call_id: z.string(),
+  output: z.union([
+    z.string(),
+    z.array(z.union([ResponseInputTextSchema, ResponseInputImageSchema, ResponseInputFileSchema])),
```
Comment (heiwen) [inline src/endpoints/conversations/schema.ts]: 1/ Why does `MessageItemUnion` have it's own name, but this thing doesn't?
2/ This should be a `discrimatedUnion` on "type", shouldn't it?

File: tsconfig.json [hunk 19]
```diff
@@ -1,7 +1,7 @@
 {
   "compilerOptions": {
     // Environment setup & latest features
-    "lib": ["ES2022", "DOM", "DOM.Iterable"],
+    "lib": ["ES2023", "DOM", "DOM.Iterable"],
     "target": "ES2022",
```
[the remark this answers is not in the document]
Comment (heiwen) [reply tsconfig.json]: @buibaoanh  Given this insights, we cannot use .toReversed()

File: src/endpoints/conversations/schema.ts [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+  type: z.literal("reasoning"),
+  id: z.string().optional(),
+  summary: z.array(ResponseSummaryTextSchema),
+  content: z.array(ResponseReasoningTextSchema).optional(),
+  encrypted_content: z.string().optional(),
+  status: ItemStatusSchema.optional(),
+});
+
+// Item: Request
+export const ResponseInputItemSchema = z.union([
```
[the remark this answers is not in the document]
Comment (heiwen) [reply src/endpoints/conversations/schema.ts]: @buibaoanh not addressed yet

File: src/endpoints/conversations/utils.ts (hunk 9 above)
Comment (heiwen) [reply src/endpoints/conversations/utils.ts]: @buibaoanh still not addressed

File: src/endpoints/conversations/storage/sql/factory.ts [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+      return this.getConversation(conversationId);
+    },
+
+    async listItems(conversationId, params) {
+      const { after, order, limit } = params;
+      let query = `SELECT * FROM conversation_items WHERE conversation_id = ${placeholder(0)}`;
+      const args: unknown[] = [conversationId];
+
+      if (after) {
+        const afterItem = await this.getItem(conversationId, after);
```
Comment (heiwen) [inline src/endpoints/conversations/storage/sql/factory.ts]: This creates a second SQL query, to avoid.

File: src/endpoints/conversations/storage/sql/factory.ts [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+          "conversations",
+          "idx_conversations_created_at",
+          ["created_at DESC"],
+          true,
+        );
+
+        await createIndex(
+          "conversation_items",
+          "idx_items_conv_id",
+          ["conversation_id", "created_at DESC", "id DESC"],
```
Comment (heiwen) [inline src/endpoints/conversations/storage/sql/factory.ts]: You’re querying getItem by (conversation_id, id) but your index starts with (conversation_id, created_at, id). This is usually still okay, but the optimal index for that exact lookup is (conversation_id, id) (or make id globally unique and just query by id).

File: src/endpoints/conversations/storage/sql/factory.ts [hunk 23]
```diff
@@ -0,0 +1,203 @@
+import type { Conversation, ConversationItem } from "../../schema";
+import type { ConversationStorage } from "../types";
+import type { QueryExecutor, DialectConfig } from "./types";
+
+interface BaseRow {
+  id: string;
+  object: string;
+  created_at: number;
+  metadata?: string | Record<string, unknown>;
```
Comment (heiwen) [inline src/endpoints/conversations/storage/sql/factory.ts]: In the future, we will need to filter conversations / items by certain metadata fields. While that's possible using JSON functions, that's slow. Greptimedb supports dynamic fields, meaning it will automatically create new columns if it hasn't seen the field before. Let's introduce an option `supportsDynamicSchema` and if that's set, metadata will not be a JSON field, but each metadata will become it's own dynamic field.

File: src/endpoints/conversations/storage/sql.ts [hunk 24]
```diff
@@ -0,0 +1,276 @@
...
+    if (cursor) {
+      const sqlParams: unknown[] = [];
+      const p = (val: unknown) => {
+        const idx = sqlParams.length;
+        sqlParams.push(val);
+        return placeholder(idx);
+      };
+
+      query = `
+          SELECT * FROM conversation_items
```
Comment (heiwen) [inline src/endpoints/conversations/storage/sql.ts]: This statement in the current form is not really cacheable.

File: package.json [hunk 25]
```diff
@@ -66,6 +71,31 @@
...
+    "./endpoints/conversations/storage/greptime": {
+      "types": "./dist/endpoints/conversations/storage/greptime.d.ts",
+      "import": "./dist/endpoints/conversations/storage/greptime.js",
+      "dev-source": "./src/endpoints/conversations/storage/greptime.ts"
+    },
+    "./endpoints/conversations/storage/mysql": {
+      "types": "./dist/endpoints/conversations/storage/mysql.d.ts",
+      "import": "./dist/endpoints/conversations/storage/mysql.js",
+      "dev-source": "./src/endpoints/conversations/storage/mysql.ts"
+    },
```
Comment (heiwen) [inline package.json]: Export everything from `storage/sql` for now, there is not much overhead.

File: src/endpoints/conversations/storage/sql.ts [hunk 26]
```diff
@@ -0,0 +1,265 @@
...
+      ]);
+    }
+  }
+
+  async createConversation(params: {
+    metadata?: Metadata;
+    items?: ResponseInputItem[];
+  }): Promise<Conversation> {
+    const { placeholder: p } = this.config;
+    const id = uuidv7();
```
Comment (heiwen) [inline src/endpoints/conversations/storage/sql.ts]: In GrepTimeDB conersation ids need to be uuidv4, since they are used as partition keys. Uuid v7 has a timestamp in the beginning, meaning the partition would only change every 2 years.

File: src/endpoints/conversations/storage/dialects/mysql.ts [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+ */
+const mapParams = (params?: unknown[]) =>
+  params?.map((p) => (p !== null && typeof p === "object" ? JSON.stringify(p) : p)) as (
+    | string
+    | number
+    | boolean
+    | null
+  )[];
+
+function isMysql2(client: any): client is Mysql2Pool {
```
Comment (heiwen) [inline src/endpoints/conversations/storage/dialects/mysql.ts]: Reminding you for time x to avoid "any". Ideally type it, if not possible use "unknown" (there are 26 "any" in this PR)

File: src/endpoints/conversations/handler.ts [hunk 28]
```diff
@@ -0,0 +1,334 @@
...
+
+    logger.debug(
+      `[conversations] added ${entities.length} items to conversation: ${conversationId}`,
+    );
+    logger.trace(
+      { requestId: ctx.requestId, conversationId, itemCount: entities.length },
+      "[storage] addItems result",
+    );
+
+    const data = entities.map((item) => toConversationItem(item));
```
Comment (heiwen) [inline src/endpoints/conversations/handler.ts]: Avoid `map` and `slice` here since they create new arrays. Do a simple `for` loop instead.

File: src/endpoints/conversations/storage/sql.ts [hunk 29]
```diff
@@ -0,0 +1,424 @@
...
+      `SELECT * FROM ${q("conversation_items")} WHERE ${q("conversation_id")} = ${p(0)}`,
+    ];
+    const args: unknown[] = [conversationId];
+    let nextIdx = 1;
+
+    if (after) {
+      // NOTE: Two-roundtrip pagination is used here for stability across all dialects (Postgres/MySQL/SQLite/Greptime).
+      // 1. First, fetch the EXACT timestamp (created_at) of the 'after' cursor ID.
+      // 2. Then, use that timestamp in a composite cursor (created_at, id) for the main query.
+      // This is more robust than a subquery as it allows for clean error handling if the cursor ID is invalid/deleted.
```
Comment (heiwen) [inline src/endpoints/conversations/storage/sql.ts]: Explain what "more robust" means in practice?

File: src/endpoints/conversations/storage/sql.ts [hunk 30]
```diff
@@ -0,0 +1,424 @@
...
+
+    const sqlParts = [`SELECT * FROM ${q("conversations")} WHERE 1=1`];
+    const args: unknown[] = [];
+    let nextIdx = 0;
+
+    // Filter by metadata
+    if (metadata && Object.keys(metadata).length > 0) {
+      for (const [key, value] of Object.entries(metadata)) {
+        // Basic sanitization for the key to prevent syntax issues if it contains quotes
+        const safeKey = key.replace(/'/g, "''");
```
Comment (heiwen) [inline src/endpoints/conversations/storage/sql.ts]: Avoid `regex` in hotpath

File: src/endpoints/conversations/storage/sql.ts [hunk 31]
```diff
@@ -0,0 +1,424 @@
...
+    await this.executor.run(
+      `DELETE FROM ${q("conversation_items")} WHERE ${q("id")} = ${p(0)} AND ${q(
+        "conversation_id",
+      )} = ${p(1)}`,
+      [itemId, conversationId],
+    );
+    return this.getConversation(conversationId);
+  }
+
+  async listItems(
```
Comment (heiwen) [inline src/endpoints/conversations/storage/sql.ts]: Codex: High: SQL pagination/order is incorrect as soon as callers supply custom item IDs, which this branch explicitly supports. src/endpoints/conversations/storage/sql.ts (line 241) assumes “sorting by ID is equivalent to sorting by time”, but src/endpoints/conversations/handler.test.ts (line 285) adds coverage that preserves caller-provided IDs. With IDs like "z" then "a", listItems(..., order: "asc") returns "a" before "z", reversing insertion/creation order and breaking after cursor semantics. This does not show up in the current SQL tests because they only use already-ordered IDs like item-1, item-2, item-3.

File: src/endpoints/conversations/storage/sql.ts [hunk 32]
```diff
@@ -0,0 +1,424 @@
...
+    );
+
+    const conversation: ConversationEntity = {
+      id,
+      created_at: now.getTime(),
+      metadata,
+    };
+
+    if (params.items?.length) {
+      await this.addItems(id, params.items);
```
Comment (heiwen) [inline src/endpoints/conversations/storage/sql.ts]: Micro optimization: addItems checks for conversation existence again. But you just created the conversation. Add an internal parameter to skip that check.

**Review by heiwen — APPROVED** [review]
Awesome work!
